statistics2-3

Mukhammed003 opened the pull request:
Выбранная командой
1. архитектура: MVVM;
2. способ вёрстки: UIKit.

Ссылка на таск трекер: https://github.com/users/zanpakt0/projects/1/views/8

Сам файл .md с декомпозицией:
[statistics_Nurmukhanov.md](https://github.com/user-attachments/files/23838091/statistics_Nurmukhanov.md)

Первоначальная оценка времени выполнения задач и реальное время, потраченное на решении каждой задачи:

## Модуль 2:
#### Верстка экрана карточки пользователя
- Настройка navigationBar для кнопки назад (est: 1 час; fact: 25 минут).
- Сверстать аватар, label для имени и доп. информаций пользователя (est: 2 часа; fact: 36 минут).
- Сверстать кнопки: 1) Перейти на сайт пользователя; 2) Коллекция NFT (est: 1 час; fact: 47 минут).
- Сверстать Alert об ошибке загрузки данных (est: 1 час; fact: 17 минут).


#### Навигация между экранами
- Настройка навигации между экранами рейтинга -> карточки пользователя (est: 1.5 часа; fact: 1 час 10 минут).
- Настройка навигации между экранами карточки пользователя -> Web View (est: 2 часа; fact: 43 минуты).
- Настройка навигации между экранами карточки пользователя -> Коллекция NFT (est: 1.5 часа; fact: 45 минут).


#### Логика экрана карточки пользователя
- Подстановка данных в аватар, label для имени и доп. информаций пользователя (est: 2 часа; fact: 11 минут).
- Дополнение структур моделей (est: 1 час; fact: 23 минуты).
- Настроить взаимосвязи между model, view и view model для паттерна MVVM (est: 2 часа; fact: 1 час).
- Логика индикатор загрузки данных (est: 1 час; fact: 35 минут).


#### Данные в карточке пользователя
- GET запрос аватарки, имени и доп. информации пользователя по ID (est: 2 часа; fact: 52 минуты).

Mukhammed003 committed:
develope(newFeature): Дополнена структура моделей

Выполнены следующие работы:
1) Добавлен файл UserCardViewData для парсинга ответа на запрос: Get user by id;
2) Обновлён endpoint в GetUsersRequest;
3) Изменено название UsersListCellViewModel на UserCellViewData для более понятного нейминга;
4) Изменена структура User (avatar: String? -> avatar: String).

Mukhammed003 committed:
develope(newFeature): Реализован метод по получению данных пользовтаеля

Выполнены следующие работы:
1) Добавлена структура GetUserCardRequest для создания запроса по получению данных пользователя;
2) Обновлён NftService с добавлением метода по загрузке данных пользователя;
3) Добавлен UserCardViewModel с методом по загрузке данных пользователя.

Mukhammed003 committed:
develope(newFeature): Настроен взаимосвязь между model, view и view model для паттерна MVVM

Выполнены следующие работы:
1) Обновлён UserCardViewController добавлением метода bindViewModel для связки с userCardViewModel.

Mukhammed003 committed:
develope(newFeature): Настроена навигация между экранами рейтинга -> карточки пользователя

Выполнены следующие работы:
1) Обновлён StatisticsViewController добавлением метода (для обработки клика по ячейке):
func tableView(_ tableView: UITableView, didSelectRowAt indexPath: IndexPath) {}

Mukhammed003 committed:
develope(newFeature): Свёрстаны нужные элементы (view) для экрана UserCardViewController

Выполнены следующие работы:
1) Свёрстан аватар, label для имени и доп. информаций пользователя;
2) Свёрстаны кнопки: 1) Перейти на сайт пользователя; 2) Коллекция NFT.

Mukhammed003 committed:
develope(newFeature): Выполнена подстановка данных в страницу UserCardViewController

Выполнены следующие работы:
1) Обновлён UserCardViewController с добавлением метода по подстановке данных в поля.

Mukhammed003 committed:
develope(newFeature): Сверстать Alert об ошибке загрузки данных

Выполнены следующие работы:
1) Обновлён UserCardViewController с добавлением алерта об ошибке.

Mukhammed003 committed:
develope(newFeature): Добавлена логика индикатора загрузки

Выполнены следующие работы в UserCardViewController:
1) Добавлен activityIndicator;
2) Скорректированы существующие view для корректного отображения данных;
3) Добавлен метод traitCollectionDidChange для правильного отображения цвета элементов при смене темы;
4) Добавлен метод showNeedViewsInScreen для показа нужных view при загрузке данных;
5) Обновлён метод bindViewModel уже с корректно отображаемой логикой всех view на экране.

Mukhammed003 committed:
develope(newFeature): Добавлен метод для отображения алерта с двумя кнопками в ErrorView

Выполнены следующие работы в UserCardViewController:
1) Обновлён Colors с добавлением forViewBackgound для установки заднего фона для WebViewVeiwController;
2) Обновлён Assets с корректировкой sortButton;
3) Добавлен метод showErrorAlertWithTwoButtons в ErrorView для показа алерта с двумя кнопками;
4) Обновлён WebViewViewController для правильного отображения в тёмной теме изменён:
1) background у view; 2) tintColor у navigationBar; 3) bottomAnchor у webView

Mukhammed003 committed:
Merge pull request #136 from zanpakt0/errorAlertWithTwoButtons

develope(newFeature): Добавлен метод для отображения алерта с двумя к…

Mukhammed003 committed:
Операция: Merge branch 'develop' of github.com:zanpakt0/iOS-FakeNFT-cohort-32-team-1 into statistics2-3

Конфликты разрешены.

Mukhammed003 committed:
develope(newFeature): Настроена навигация между экранами карточки пользователя -> Web View

Выполнены следующие работы:
1) В UserCardViewController добавлен метод goToUserWebsiteButtonClicked по открытию website юзера(через WebView);
2) В UserCardViewControlelr и StatisticsViewController обновлён метод по показу алерта об ошибке (showErrorAlert).

Mukhammed003 committed:
develope(newFeature): Настроена навигации между экранами карточки пользователя -> Коллекция NFT

Выполнены следующие работы:
1) В UserCardViewController добавлен метод nftCollectionButtonClicked по открытию страницы(UserCollectionViewController) с коллекцией юзера;
2) Добавлены файлы UserCollectionViewCell, UserCollectionViewController, UserCollectionVIewModel для отображения коллекции nft юзера;
3) Добавлен viewWillAppear в StatisticsViewController для подгрузки юзеров при повторном входе в StatisticsViewController;
4) Добавлены маркировки.

Mukhammed003 committed:
develope(newFeature): Обновил файл с декомпозицией эпика "Статистика"

Mono1it commented on FakeNFT/Scenes /Statistics/NeedStructures/UserCardViewData.swift:
```diff
@@ -0,0 +1,25 @@
+import Foundation
+
+// MARK: - Structure of user's info (Need for user card page)
+final class UserCardViewData {
```
Возможно стоит сделать это Структурой.

Mono1it commented on FakeNFT/Scenes /Statistics/NeedStructures/UserCellViewData.swift:
```diff
@@ -7,8 +7,8 @@
-// MARK: - Structure of user's info (Need for tableView in statistics page)
-final class UsersListCellViewModel: Equatable {
+// MARK: - Structure of users info (Need for tableView in statistics page)
+final class UserCellViewData: Equatable {
```
Как я понял UserCellViewData — это простой контейнер данных без состояния.
И по хорошему он должен быть структурой.

Mono1it commented on FakeNFT/Scenes /Statistics/Requests/GetUsersRequest.swift:
```diff
@@ -12,7 +12,13 @@ struct GetUsersRequest: NetworkRequest {
     // MARK: - Endpoint
     var endpoint: URL? {
-        URL(string: "https://d5dn3j2ouj72b0ejucbl.apigw.yandexcloud.net/api/v1/users?page=\(page)&size=\(size)")
+        var components = URLComponents(string: RequestConstants.baseURL)
```
Супер, хорошо что используешь URLComponents

Mono1it commented on FakeNFT/Scenes /Statistics/UserCard/UserCardViewController.swift:
```diff
@@ -0,0 +1,343 @@
+        let exampleImage = UIImage(systemName: "person.crop.circle.fill")
+        let userAvatarImageView = UIImageView(image: exampleImage)
+        userAvatarImageView.clipsToBounds = true
+        userAvatarImageView.layer.cornerRadius = 35
```
Можно добавить данное значение в UserCardLayout

Mono1it commented on FakeNFT/Scenes /Statistics/UserCard/UserCardViewController.swift:
```diff
@@ -0,0 +1,343 @@
+        arrowImage.translatesAutoresizingMaskIntoConstraints = false
+        
+        NSLayoutConstraint.activate([
+            stack.leadingAnchor.constraint(equalTo: nftCollectionButton.leadingAnchor, constant: 16),
```
Также данные констрейты можно добавить в UserCardLayout

Mono1it commented on FakeNFT/Scenes /Statistics/UserCard/UserCardViewController.swift:
```diff
@@ -0,0 +1,343 @@
+        let webView = WebViewViewController(urlString: webSiteURL.absoluteString)
+        
+        let transition = CATransition()
+        transition.duration = 0.01
```
Тут лучше использовать отдельную переменную
```
private let timeOfDuration = 0.01

transition.duration = timeOfDuration
```

Mono1it commented on FakeNFT/Scenes /Statistics/UserCard/UserCardViewController.swift:
```diff
@@ -0,0 +1,343 @@
+        let userCollectionViewController = UserCollectionViewController(viewModel: userCollectionViewModel)
+        
+        let transition = CATransition()
+        transition.duration = 0.01
```
Тут лучше использовать отдельную переменную
```
private let timeOfDuration = 0.01

transition.duration = timeOfDuration
```

Mono1it commented on FakeNFT/Scenes /Statistics/UserCard/UserCardViewController.swift:
```diff
@@ -0,0 +1,343 @@
+    }
+    
+    private func insertUserDataIntoFields(userData: UserCardViewData) {
+        let processor = RoundCornerImageProcessor(cornerRadius: 35)
```
Можно добавить данное значение в UserCardLayout

Mono1it commented on FakeNFT/Scenes /Statistics/UserCard/UserCardViewModel.swift:
```diff
@@ -0,0 +1,67 @@
+        
+        currentTask = serviceAssembly.nftService.getUserCard(id: id) {[weak self] result in
+            DispatchQueue.main.async {
+                guard let self = self else { return }
```
Тут можно не писать `self = self`, компилятор примет: `guard let self else { return }`

Mono1it commented on FakeNFT/Scenes /Statistics/UserCollection/UserCollectionViewCell.swift:
```diff
@@ -0,0 +1,12 @@
+
+import UIKit
+
+final class UserCollectionViewCell: UICollectionViewCell {
```
Пустая ячейка?

Mono1it commented on FakeNFT/Scenes /Statistics/NeedStructures/UserCardViewData.swift:
```diff
@@ -0,0 +1,25 @@
+//  FakeNFT
+//
+//  Created by Muhammed Nurmukhanov on 26.11.2025.
+//
```
Удаляй ненужные данные

Mono1it commented on FakeNFT/Scenes /Statistics/NeedStructures/UserCardViewData.swift:
```diff
@@ -0,0 +1,25 @@
+        self.website = URL(string: user.website)
+        self.nfts = user.nfts
+    }
+}
```
Возможно стоит сделать это Структурой.

Mono1it commented on FakeNFT/Scenes /Statistics/NeedStructures/UserCellViewData.swift:
Как я понял UserCellViewData — это простой контейнер данных без состояния.
И по хорошему он должен быть структурой.

Mono1it commented on FakeNFT/Scenes /Statistics/UserCard/UserCardViewController.swift:
```diff
@@ -0,0 +1,343 @@
+    private var cancellables = Set<AnyCancellable>()
+    
+    // MARK: - Views (elements)
+    lazy var activityIndicator: UIActivityIndicatorView = {
```
Можно сделать свойство private, чтобы не "светить им наружу"

Mukhammed003 replied:
Не могу сделать так как наследуюсь от LoadingView, и он требует чтобы он был публичным.

Mono1it commented on FakeNFT/Scenes /Statistics/UserCard/UserCardViewController.swift:
```diff
@@ -0,0 +1,343 @@
+        goToUserWebsiteButton.layer.borderWidth = 1
+        goToUserWebsiteButton.layer.borderColor = UIColor.segmentActive.cgColor
+        
+        goToUserWebsiteButton.layer.cornerRadius = 16
```
Можно добавить данное значение в UserCardLayout

Mono1it commented on FakeNFT/Scenes /Statistics/UserCollection/UserCollectionViewCell.swift:
```diff
@@ -0,0 +1,12 @@
+//  Created by Muhammed Nurmukhanov on 30.11.2025.
+//
+
+import UIKit
+
+final class UserCollectionViewCell: UICollectionViewCell {
+    
+}
```
Пустая ячейка?

Mukhammed003 replied:
Это заготовка для 3 модуля. Оно нужно чтобы выполнить навигацию с карточки пользователя в коллекцию Nft пользователя.

Mono1it commented on FakeNFT/Scenes /Statistics/UserList/StatisticsTableViewCell.swift:
```diff
@@ -103,7 +104,7 @@ final class StatisticsTableViewCell: UITableViewCell {
-    func configure(numberingOfCell: Int, with usersInfo: UsersListCellViewModel) {
+    func configure(numberingOfCell: Int, with usersInfo: UserCellViewData) {
         self.numberingLabel.text = "\(numberingOfCell)"
         let processor = RoundCornerImageProcessor(cornerRadius: 14)
```
Так же можно добавить в StatisticsTableViewCellLayout

Mono1it reviewed (commented):
Насчет структур не очень уверен, думаю стоит обсудить это вопрос с наставником или ревьювер практикума подскажет

Mukhammed003 committed:
fix(bugs): Поправил по замечаниям от Ильи(сокомандник)

Выполнены следующие работы:
1) Во всех файлах были очищены не нужные комменты;
2) UserCardViewData, UserCellViewData были переопределены как структуры;
3) Во всех файлах(где есть магические числа) по возможности были убраны магические числа (записаны как константы).
